src/content/post/fastrace-a-modern-approach-to-distributed-tracing-in-rust.mdx

```diff
@@ -0,0 +1,374 @@
+        user_id = 42, 
+        request_id = "abcd1234"
+    );
+    
+    // Enter the span (activates it for the current execution context)
+    let _guard = span.enter();
+    
+    // Log within the span context
+    tracing::info!("Starting request processing");
+    
+    process_data();
+    
+    tracing::info!("Finished processing request");
+}
+```
+
+For instrumenting functions, `tokio-rs/tracing` provides attribute macros:
+
+```rust
+use tracing::instrument;
+
+#[instrument(skip(password), fields(user_id = user.id))]
+async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
+    info!("Authenticating user {}", user.id);
+    // ...more code...
+}
+```
+
+## The Challenges with `tokio-rs/tracing`
+
+While powerful, `tokio-rs/tracing` comes with several significant challenges:
+
+### 1. Ecosystem Fragmentation
+
+By introducing its own logging macros, `tokio-rs/tracing` creates a division with code using the standard `log` crate:
+
+```rust
+// Using log crate
+log::info!("Starting operation");
+
+// Using tracing crate (different syntax)
+tracing::info!("Starting operation");
+```
+
+This fragmentation is particularly problematic for library authors. When creating a library, authors face a difficult choice:
+
+1. Use the `log` crate for compatibility with the broader ecosystem
+2. Use `tokio-rs/tracing` for better observability features
+3. Implement dual support for both systems (increasing maintenance burden)
+
+    This often leads to complex feature-flag systems and wrapper macros:
+
+    ```rust
+    macro_rules! info {
+        ($($arg:tt)*) => {
+            if cfg!(feature = "tracing") {
+                tracing::info!($($arg)*);
+            } else {
+                log::info!($($arg)*);
+            }
+        }
+    }
+    ```
+
+Many libraries choose the first option for simplicity, but miss out on the benefits of tracing. This creates a situation where applications have inconsistent instrumentation across their dependency tree.
+
+### 2. Performance Impact for Libraries
+
+Library authors are particularly sensitive to performance overhead, as their code may be called in tight loops or performance-critical paths. `tokio-rs/tracing`'s overhead can be substantial when enabled, which creates a dilemma:
+
+1. Always enable tracing (and impose overhead on all users)
+2. Don't instrument at all (and lose observability)
+3. Create a complex feature-flag system (increasing maintenance burden)
+
+    This is a common pattern in libraries using `tokio-rs/tracing`:
+
+    ```rust
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip(password), fields(user_id = user.id)))]
+    async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
+        // ...more code...
+    }
+    ```
+
+With `tokio-rs/tracing`, there's no clean way to have zero-cost tracing when disabled. This makes library authors reluctant to add instrumentation to performance-sensitive code paths.
```

Comment:
fastrace has an `enabled` flag. But tracing doesn't have any similar mechanism?

I don't get it quite clearly what "no clean way to have zero-cost tracing when disabled" means, especially what "when (tracing) disabled" means, or how to.